HTTPRequest/HTTPRequest.DotNet.cs

```diff
@@ -0,0 +1,192 @@
+        {
+            httpRequest.ContentType = contentType;
+        }
+
+        // Set cookies if provided
+        if (cookies != null)
+        {
+            httpRequest.CookieContainer = cookies;
+        }
+
+        // Setup private properties and fire off the request
+        this.cb = cb;
+        this.cookies = cookies;
+        this.request = httpRequest;
+
+        // Initiate response
+        if (postData == null)
+        {
+            ReadResponseData(cb);
+            return;
+        } else
+        {
+            WritePostData(postData, (Exception requestError) => {
+                if (requestError != null)
+                {
+                    cb(requestError, null);
+                    return;
+                }
+
+                // Process the response
+                ReadResponseData(cb);
+            });
+            return;
+        }
+    }
+
+
+    // Write post data to request buffer
+    private void WritePostData(byte[] postData, Action<Exception> cb)
+    {
+        request.BeginGetRequestStream((IAsyncResult callbackResult) => {
+            try
+            {
+                Stream postStream = request.EndGetRequestStream(callbackResult);
+                postStream.Write(postData, 0, postData.Length);
+                postStream.Close();
+            }
+            catch (Exception error)
+            {
+                cb(error);
+                return;
+            }
+            cb(null);
+        }, null);
+    }
+
+    // Read response data from request buffer
+    private void ReadResponseData(Action<Exception, string> cb)
+    {
+        // Begin waiting for a response
+        request.BeginGetResponse(new AsyncCallback((IAsyncResult callbackResult) => {
+            // Cleanup timeout
+            timeoutTimer.Dispose();
+            timeoutTimer = null;
+
+            // Process response
+            string responseString = null;
+            try
+            {
+                HttpWebResponse response = (HttpWebResponse)request.EndGetResponse(callbackResult);
+
+                using (StreamReader httpWebStreamReader = new StreamReader(response.GetResponseStream()))
+                {
+                    responseString = httpWebStreamReader.ReadToEnd();
+                }
+
+                response.Close();
+            }
+            catch (Exception error)
+            {
+                cb(error, null);
+                return;
+            }
+
+            cb(null, responseString);
+        }), null);
+    }
+
+
+    // Abort request
+    public void Abort()
+    {
+        if (request == null)
+        {
+            return;
+        }
+
+        HttpWebRequest _request = request;
+        request = null;
+
+        _request.Abort();
+        timeoutTimer.Dispose();
+        timeoutTimer = null;
+    }
+
+
+    // Create GET request and return it
+    public static HTTPRequest Get(string url, Dictionary<string, string> headers, CookieContainer cookies, Action<Exception, string> cb)

```

Comment:
At some point I would like to pull this out, I can see what you are doing here, but something over arching would make this much easier to test in the future.